contracts/QuestFactory.sol

```diff
@@ -369,6 +368,9 @@ contract QuestFactory is Initializable, LegacyStorage, OwnableRoles, IQuestFacto
             (address, address, string, string)
         );
         Quest storage quest = quests[questId_];
+
+        if(tx.origin != msg.sender && tx.origin != address(this) && tx.origin != quest.questAddress) revert txOriginMismatch();
```

Comment:
For the `address(this)` case, we should be able to split this into an internal helper `_claimOptimized` and an external to get around needing to do this extra check. Also same question for what case is the `tx.origin` actually the `questAddress`.